File: src/Frontend/factutrust-web/src/app/features/settings/numbering/numbering-format-renderer.ts
import {
  BLOCK_TYPE_LABELS,
  isDocumentNumberBlock,
  NumberingBlock,
  NumberingBlockType
} from './models/numbering.models';

export const MAX_RENDERED_LENGTH = 50;

export interface RenderValidationResult {
  valid: boolean;
  error?: string;
}

function padSequence(sequence: number, digits: number): string {
  return sequence.toString().padStart(digits, '0');
}

function renderBlock(
  block: NumberingBlock,
  sequence: number,
  referenceDate: Date,
  freeTextOverride?: string
): string | null {
  switch (block.type) {
    case NumberingBlockType.FreeText:
      return (freeTextOverride ?? block.value ?? '').trim().toUpperCase();
    case NumberingBlockType.Separator:
      return block.value ?? '-';
    case NumberingBlockType.DocumentNumber:
      return sequence.toString();
    case NumberingBlockType.DocumentNumberPadded3:
      return padSequence(sequence, 3);
    case NumberingBlockType.DocumentNumberPadded4:
      return padSequence(sequence, 4);
    case NumberingBlockType.DocumentNumberPadded5:
      return padSequence(sequence, 5);
    case NumberingBlockType.DocumentNumberPadded6:
      return padSequence(sequence, 6);
    case NumberingBlockType.Day:
      return referenceDate.getDate().toString().padStart(2, '0');
    case NumberingBlockType.Month:
      return (referenceDate.getMonth() + 1).toString().padStart(2, '0');
    case NumberingBlockType.Year4:
      return referenceDate.getFullYear().toString();
    case NumberingBlockType.Year2:
      return (referenceDate.getFullYear() % 100).toString().padStart(2, '0');
    default:
      return null;
  }
}

export function render(
  blocks: NumberingBlock[],
  sequence: number,
  referenceDate: Date,
  freeTextOverride?: string
): { value: string; error?: string } {
  const validation = validateBlocks(blocks);
  if (!validation.valid) {
    return { value: '', error: validation.error };
  }

  const ordered = [...blocks].sort((a, b) => a.order - b.order);
  const parts: string[] = [];

  for (const block of ordered) {
    const part = renderBlock(block, sequence, referenceDate, freeTextOverride);
    if (part === null) {
      return { value: '', error: `Type de bloc inconnu: ${block.type}` };
    }
    parts.push(part);
  }

  const rendered = parts.join('');
  if (rendered.length > MAX_RENDERED_LENGTH) {
    return {
      value: '',
      error: `Le numéro généré dépasse ${MAX_RENDERED_LENGTH} caractères.`
    };
  }

  return { value: rendered };
}

export function validateBlocks(blocks: NumberingBlock[]): RenderValidationResult {
  if (!blocks.length) {
    return { valid: false, error: 'Le format de numérotation est vide.' };
  }

  if (!blocks.some((b) => isDocumentNumberBlock(b.type))) {
    return {
      valid: false,
      error: 'Le format doit contenir au moins un bloc numéro de document.'
    };
  }

  for (const block of blocks) {
    if (block.type === NumberingBlockType.FreeText) {
      if (!block.value || !block.value.trim()) {
        return { valid: false, error: 'Le texte libre est obligatoire.' };
      }
      if (block.value.trim().length > 20) {
        return { valid: false, error: 'Le texte libre ne peut pas dépasser 20 caractères.' };
      }
    }

    if (block.type === NumberingBlockType.Separator) {
      if (block.value !== '-' && block.value !== '/') {
        return { valid: false, error: 'Le séparateur doit être « - » ou « / ».' };
      }
    }
  }

  return { valid: true };
}

export function blockDisplayLabel(block: NumberingBlock): string {
  return block.label ?? BLOCK_TYPE_LABELS[block.type];
}

File: src/Backend/FactuTrust.Domain/Services/NumberingFormatRenderer.cs
using FactuTrust.Domain.Common;
using FactuTrust.Domain.Enums;
using FactuTrust.Domain.ValueObjects;

namespace FactuTrust.Domain.Services;

public static class NumberingFormatRenderer
{
    public const int MaxRenderedLength = 50;

    public sealed record RenderContext(
        int Sequence,
        DateTime ReferenceDate,
        string? FreeTextOverride = null);

    public static Result<string> Render(
        IReadOnlyList<NumberingFormatBlock> blocks,
        RenderContext context)
    {
        if (blocks.Count == 0)
            return Result.Failure<string>(Error.Validation("Format", "Le format de numerotation est vide."));

        if (!blocks.Any(b => b.Type.IsDocumentNumberBlock()))
            return Result.Failure<string>(Error.Validation("Format",
                "Le format doit contenir au moins un bloc numero de document."));

        var ordered = blocks.OrderBy(b => b.Order).ToList();
        var parts = new List<string>(ordered.Count);

        foreach (var block in ordered)
        {
            var partResult = RenderBlock(block, context);
            if (partResult.IsFailure)
                return partResult;
            parts.Add(partResult.Value);
        }

        var rendered = string.Concat(parts);
        if (rendered.Length > MaxRenderedLength)
            return Result.Failure<string>(Error.Validation("Format",
                $"Le numero genere depasse {MaxRenderedLength} caracteres."));

        return Result.Success(rendered);
    }

    public static Result ValidateBlocks(IReadOnlyList<NumberingFormatBlock> blocks)
    {
        if (blocks.Count == 0)
            return Result.Failure(Error.Validation("Format", "Le format de numerotation est vide."));

        if (!blocks.Any(b => b.Type.IsDocumentNumberBlock()))
            return Result.Failure(Error.Validation("Format",
                "Le format doit contenir au moins un bloc numero de document."));

        foreach (var block in blocks)
        {
            switch (block.Type)
            {
                case NumberingBlockType.FreeText:
                    if (string.IsNullOrWhiteSpace(block.Value))
                        return Result.Failure(Error.Validation("FreeText", "Le texte libre est obligatoire."));
                    if (block.Value.Trim().Length > 20)
                        return Result.Failure(Error.Validation("FreeText",
                            "Le texte libre ne peut pas depasser 20 caracteres."));
                    break;
                case NumberingBlockType.Separator:
                    if (block.Value is not ("-" or "/"))
                        return Result.Failure(Error.Validation("Separator",
                            "Le separateur doit etre '-' ou '/'."));
                    break;
            }
        }

        return Result.Success();
    }

    private static Result<string> RenderBlock(NumberingFormatBlock block, RenderContext context)
    {
        return block.Type switch
        {
            NumberingBlockType.FreeText => Result.Success(
                (context.FreeTextOverride ?? block.Value ?? string.Empty).Trim().ToUpperInvariant()),
            NumberingBlockType.Separator => Result.Success(block.Value ?? "-"),
            NumberingBlockType.DocumentNumber => Result.Success(context.Sequence.ToString()),
            NumberingBlockType.DocumentNumberPadded3 => Result.Success(context.Sequence.ToString("D3")),
            NumberingBlockType.DocumentNumberPadded4 => Result.Success(context.Sequence.ToString("D4")),
            NumberingBlockType.DocumentNumberPadded5 => Result.Success(context.Sequence.ToString("D5")),
            NumberingBlockType.DocumentNumberPadded6 => Result.Success(context.Sequence.ToString("D6")),
            NumberingBlockType.Day => Result.Success(context.ReferenceDate.Day.ToString("D2")),
            NumberingBlockType.Month => Result.Success(context.ReferenceDate.Month.ToString("D2")),
            NumberingBlockType.Year4 => Result.Success(context.ReferenceDate.Year.ToString("D4")),
            NumberingBlockType.Year2 => Result.Success((context.ReferenceDate.Year % 100).ToString("D2")),
            _ => Result.Failure<string>(Error.Validation("BlockType", $"Type de bloc inconnu: {block.Type}"))
        };
    }
}